### ISSUE TITLE
Allow where clause predicates that contain no type parameters

### ISSUE THREAD

**Issue by erickt** [rust-lang/rust#21203]
Where clauses right now require predicates to contain type parameters, as in:

```rust
trait Foo {
    fn foo(&self);
}

struct Bar<T> {
    x: T,
}

impl<T> Foo for Bar<T> where T: Foo {
    fn foo(&self) {
        self.x.foo()
    }
}
```

But we cannot add constraints on non-generic types:

```rust
struct Baz { x: i32 }

impl Foo for Baz where i32: Foo {
    fn foo(&self) {
        self.x.foo()
    }
}
```

Will error out with:

```
test.rs:15:1: 19:2 error: cannot bound type `i32`, where clause bounds may only be attached to types involving type parameters
test.rs:15 impl Foo for Baz where i32: Foo {
test.rs:16     fn foo(&self) {
test.rs:17         self.x.foo()
test.rs:18     }
test.rs:19 }
```

In hand written code this usually won't come up, but this would be really helpful for a code generator like `macro_rules!` or `#[derive]`, especially when dealing with associated types. Currently the only way to know if a type is generic or not is to manually walk it and see if any of the type paths start with the same name as a type parameter. If so, add that type to the predicate list. Instead, it would be much simpler if we could just add all the types listed in a field or enum to the predicate list and let the type checker report an error if one of those types doesn't implement the trait.

cc @nikomatsakis, @jroesch

**Comment by nikomatsakis** [comment]
So I think this is a good reason to just permit where clauses like this. Ideally we'll report errors at the callee, though it's easiest to report errors at the caller.

**Comment by nikomatsakis** [comment]
@erickt you seem to have disappeared from IRC, but the point about privacy is a good one -- I feel like the privacy rules are being too strict here.

**Comment by nikomatsakis** [comment]
I'm not so sure about the privacy thing any more. After some conversation with @aturon we felt like the current rules might be best.

**Comment by Gankro** [comment]
Should this be closed as wontfix/intended, then?

**Comment by jroesch** [comment]
I lost track of this, did anyone remove the check that ensures bound types include a type parameter? I remember we had talked about shuffling around some error reporting as well, but its been too many days since I've thought about this.

**Comment by huonw** [comment]
@Gankro I think the privacy rules are somewhat orthogonal to allowing non-generic where clauses (although the motivating use case is definitely undermined by the current privacy rules).

**Comment by steveklabnik** [comment]
Triage: this now works. Full reproduction:

```
trait Foo {
    fn foo(&self);
}

impl Foo for i32 {
    fn foo(&self) { println!("success {}", self); }
}

struct Baz { x: i32 }

impl Foo for Baz where i32: Foo {
    fn foo(&self) {
        self.x.foo()
    }
}

fn main() {
    let b = Baz { x: 5 };
    b.foo();
}

```

**Comment by nikomatsakis** [comment]
Huh. When did that happen? I know there have been some pending PRs about this, but there were some thorny issues, and I didn't realize that we'd made any actual changes here. @arielb1, do you know?

**Comment by arielb1** [comment]
What? I didn't implement this?

**Comment by mitaa** [comment]
I think this is the relevant commit: https://github.com/rust-lang/rust/commit/dbf994bbaff214e6441a8e97df180db7b29e7189

**Comment by nikomatsakis** [comment]
Yeah I was wondering if this was somehow fallout from that change. That's a bit unexpected though.

**Comment by nikomatsakis** [comment]
So we do need a solution that addresses this example raised by @arielb1 :+1: 

```rust
#[no_mangle]
pub fn bogus<'a>(x: &'a mut ()) where &'a mut (): Clone {
    <&'a mut ()>::clone(&x);
}
```

The key point in this example is that this where clause could never hold -- but because it's only parametric over a lifetime, we will try to trans it, and find that there is no existing impl. This could either fail in typeck -- but we'd have to search for impls using an erased parameter -- or else fail in trans -- where the absence of an impl could cause us to emit a trap or something like that, on the premise that the fn should not in fact be callable.

Basically in trans we can just check if we know statically that the fn could never be called and generate a trap or what have you. Or just don't translate it. Or something. This seems similar to the logic that we use to screen things out of vtables.

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by nrc** [comment]
triage: P-medium

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by brson** [comment]
P-low?

**Comment by steveklabnik** [comment]
Triage: @nikomatsakis 's sample compiles, but if you try to use it, can never work. Aka, this is still an issue, as far as I can tell.

**Comment by vandenheuvel** [comment]
https://github.com/rust-lang/rust/issues/75961 is a duplicate of @nikomatsakis 's example. That issue contains more recent diagnostics: the example no longer compiles.

**Comment by vandenheuvel** [comment]
The original issue now compiles with the trivial bounds feature, see https://github.com/rust-lang/rust/issues/48214.
